### DISCUSSION TITLE
How do I intercept request, extract relavent info and forward the original request with a custom object?

### DISCUSSION THREAD

**Discussion by cometw** [tokio-rs/axum#2101] [Q&A] [answered]
I have a `tower::Service` impl with a nested `inner` (also impls `tower::Service`)

``` rust
pub(crate) struct CustomSession {
    inner: SessionAuth<...<...<...<FetchSession>>>,
}

impl Service<ApiRequest> for CustomSession {
    type Response = UserSession;

    type Error = FetchSessionError;

    type Future = BoxFuture<Self::Response, Self::Error>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        ...
    }

    fn call(&mut self, api_req: ApiRequest) -> Self::Future {
        self.inner.call(api_req)
    }
```

`ApiRequest` essentially looks like this

``` rust
#[derive(Debug, Clone)]
pub(crate) struct ApiRequest {
    pub(crate) scheme: Option<String>,
    pub(crate) remote_addr: Option<String>,
    pub(crate) user_agent: Option<String>,
    pub(crate) api_key: Option<String>,
}
```

Based on this req, each layer in

``` rust
    inner: SessionAuth<...<...<...<FetchSession>>>,
```

contacts a backend database and either retrieves or verifies some information and finally return a `UserSession` object.

I want to be able to use this `UserSession` in route handlers, like this

``` rust
pub(crate) async fn user_fetch(
    fetch_svc: State<Fetch>, // Service to fetch user info.
    UserSession(session): UserSession, // <--- This here
    Json(user_id): Json<UserId>, // Unique user Id.
) -> impl IntoResponse {
    let user = fetch_svc
        ... // arc/mutex magic
        ... 
        .call(IdentifiableUser { user_id })
        ...
        ...
        .await?;

    Ok(user)
}
```

Accesssing this path without `UserSession` should return an error.

Is this possible? And if so, I am not sure what to look for.

I think I'll need to refactor `CustomSession` into a `Layer`,
perhaps also change the service type to take in `Request<Body>` and extract info to generate `ApiRequest` in the call method. 

Also not sure how I'd extract `api_key` from query params. If it were a handler, I could've done
``` rust
Query(api_key): Query<ApiKey>
```
where `ApiKey` is a wrapper that derives `Deserialize`.

**Comment by davidpdrsn** [comment, marked as answer]
In your middleware you’ll wanna add additional data as request extensions and extract those in the handler. See https://docs.rs/axum/latest/axum/middleware/index.html#passing-state-from-middleware-to-handlers

**Reply by cometw** [reply]
Ah so I should insert my `ApiRequest` like this
``` rust
req.extensions_mut().insert(api_request);
```
Another question if you don't mind - 

`middleware::from_fn` would work here, but would it be possible to get access to `mut req: Request<B>, next: Next<B>` in my service's `call` method?

If so, how would I deserialize `ApiKey` in there? I imagine something like
``` rust
let Query(api_key) = req.<method>();
```

**Reply by davidpdrsn** [reply]
You already get the request in call and “next” is the inner service your service is wrapping.

**Reply by cometw** [reply]
Oh that makes sense, so a generic `impl <S, B>` with `Service` constrained on `Request<B>` would work. Thanks for the answer!
